// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: notification/v1/notification.proto

package notificationsv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Notification with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Notification) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Notification with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in NotificationMultiError, or
// nil if none found.
func (m *Notification) ValidateAll() error {
	return m.validate(true)
}

func (m *Notification) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Type

	// no validation rules for State

	if all {
		switch v := interface{}(m.GetAction()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "Action",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "Action",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetAction()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NotificationValidationError{
				field:  "Action",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetCreated()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "Created",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "Created",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCreated()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NotificationValidationError{
				field:  "Created",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, NotificationValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NotificationValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return NotificationMultiError(errors)
	}

	return nil
}

// NotificationMultiError is an error wrapping multiple validation errors
// returned by Notification.ValidateAll() if the designated constraints aren't met.
type NotificationMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NotificationMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NotificationMultiError) AllErrors() []error { return m }

// NotificationValidationError is the validation error returned by
// Notification.Validate if the designated constraints aren't met.
type NotificationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NotificationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NotificationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NotificationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NotificationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NotificationValidationError) ErrorName() string { return "NotificationValidationError" }

// Error satisfies the builtin error interface
func (e NotificationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotification.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NotificationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NotificationValidationError{}

// Validate checks the field values on NotificationFilter with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *NotificationFilter) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NotificationFilter with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// NotificationFilterMultiError, or nil if none found.
func (m *NotificationFilter) ValidateAll() error {
	return m.validate(true)
}

func (m *NotificationFilter) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return NotificationFilterMultiError(errors)
	}

	return nil
}

// NotificationFilterMultiError is an error wrapping multiple validation errors
// returned by NotificationFilter.ValidateAll() if the designated constraints
// aren't met.
type NotificationFilterMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NotificationFilterMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NotificationFilterMultiError) AllErrors() []error { return m }

// NotificationFilterValidationError is the validation error returned by
// NotificationFilter.Validate if the designated constraints aren't met.
type NotificationFilterValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NotificationFilterValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NotificationFilterValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NotificationFilterValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NotificationFilterValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NotificationFilterValidationError) ErrorName() string {
	return "NotificationFilterValidationError"
}

// Error satisfies the builtin error interface
func (e NotificationFilterValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotificationFilter.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NotificationFilterValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NotificationFilterValidationError{}

// Validate checks the field values on NotificationsFilter with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *NotificationsFilter) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NotificationsFilter with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// NotificationsFilterMultiError, or nil if none found.
func (m *NotificationsFilter) ValidateAll() error {
	return m.validate(true)
}

func (m *NotificationsFilter) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.User != nil {

		if all {
			switch v := interface{}(m.GetUser()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, NotificationsFilterValidationError{
						field:  "User",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, NotificationsFilterValidationError{
						field:  "User",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NotificationsFilterValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if m.Params != nil {

		if all {
			switch v := interface{}(m.GetParams()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, NotificationsFilterValidationError{
						field:  "Params",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, NotificationsFilterValidationError{
						field:  "Params",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetParams()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return NotificationsFilterValidationError{
					field:  "Params",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return NotificationsFilterMultiError(errors)
	}

	return nil
}

// NotificationsFilterMultiError is an error wrapping multiple validation
// errors returned by NotificationsFilter.ValidateAll() if the designated
// constraints aren't met.
type NotificationsFilterMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NotificationsFilterMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NotificationsFilterMultiError) AllErrors() []error { return m }

// NotificationsFilterValidationError is the validation error returned by
// NotificationsFilter.Validate if the designated constraints aren't met.
type NotificationsFilterValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NotificationsFilterValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NotificationsFilterValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NotificationsFilterValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NotificationsFilterValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NotificationsFilterValidationError) ErrorName() string {
	return "NotificationsFilterValidationError"
}

// Error satisfies the builtin error interface
func (e NotificationsFilterValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotificationsFilter.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NotificationsFilterValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NotificationsFilterValidationError{}

// Validate checks the field values on AddNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *AddNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AddNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// AddNotificationRequestMultiError, or nil if none found.
func (m *AddNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *AddNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, AddNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, AddNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return AddNotificationRequestValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return AddNotificationRequestMultiError(errors)
	}

	return nil
}

// AddNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by AddNotificationRequest.ValidateAll() if the designated
// constraints aren't met.
type AddNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AddNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AddNotificationRequestMultiError) AllErrors() []error { return m }

// AddNotificationRequestValidationError is the validation error returned by
// AddNotificationRequest.Validate if the designated constraints aren't met.
type AddNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AddNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AddNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AddNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AddNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AddNotificationRequestValidationError) ErrorName() string {
	return "AddNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e AddNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAddNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AddNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AddNotificationRequestValidationError{}

// Validate checks the field values on AddNotificationResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *AddNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AddNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// AddNotificationResponseMultiError, or nil if none found.
func (m *AddNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *AddNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Added

	if len(errors) > 0 {
		return AddNotificationResponseMultiError(errors)
	}

	return nil
}

// AddNotificationResponseMultiError is an error wrapping multiple validation
// errors returned by AddNotificationResponse.ValidateAll() if the designated
// constraints aren't met.
type AddNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AddNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AddNotificationResponseMultiError) AllErrors() []error { return m }

// AddNotificationResponseValidationError is the validation error returned by
// AddNotificationResponse.Validate if the designated constraints aren't met.
type AddNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AddNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AddNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AddNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AddNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AddNotificationResponseValidationError) ErrorName() string {
	return "AddNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e AddNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAddNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AddNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AddNotificationResponseValidationError{}

// Validate checks the field values on EditNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *EditNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on EditNotificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// EditNotificationRequestMultiError, or nil if none found.
func (m *EditNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *EditNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, EditNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, EditNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return EditNotificationRequestValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return EditNotificationRequestMultiError(errors)
	}

	return nil
}

// EditNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by EditNotificationRequest.ValidateAll() if the designated
// constraints aren't met.
type EditNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m EditNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m EditNotificationRequestMultiError) AllErrors() []error { return m }

// EditNotificationRequestValidationError is the validation error returned by
// EditNotificationRequest.Validate if the designated constraints aren't met.
type EditNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EditNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EditNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EditNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EditNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EditNotificationRequestValidationError) ErrorName() string {
	return "EditNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e EditNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEditNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EditNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EditNotificationRequestValidationError{}

// Validate checks the field values on EditNotificationResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *EditNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on EditNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// EditNotificationResponseMultiError, or nil if none found.
func (m *EditNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *EditNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Edited

	if len(errors) > 0 {
		return EditNotificationResponseMultiError(errors)
	}

	return nil
}

// EditNotificationResponseMultiError is an error wrapping multiple validation
// errors returned by EditNotificationResponse.ValidateAll() if the designated
// constraints aren't met.
type EditNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m EditNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m EditNotificationResponseMultiError) AllErrors() []error { return m }

// EditNotificationResponseValidationError is the validation error returned by
// EditNotificationResponse.Validate if the designated constraints aren't met.
type EditNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EditNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EditNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EditNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EditNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EditNotificationResponseValidationError) ErrorName() string {
	return "EditNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e EditNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEditNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EditNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EditNotificationResponseValidationError{}

// Validate checks the field values on GetNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetNotificationRequestMultiError, or nil if none found.
func (m *GetNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFilter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetNotificationRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetNotificationRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFilter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetNotificationRequestValidationError{
				field:  "Filter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetNotificationRequestMultiError(errors)
	}

	return nil
}

// GetNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by GetNotificationRequest.ValidateAll() if the designated
// constraints aren't met.
type GetNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetNotificationRequestMultiError) AllErrors() []error { return m }

// GetNotificationRequestValidationError is the validation error returned by
// GetNotificationRequest.Validate if the designated constraints aren't met.
type GetNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetNotificationRequestValidationError) ErrorName() string {
	return "GetNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetNotificationRequestValidationError{}

// Validate checks the field values on GetNotificationResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetNotificationResponseMultiError, or nil if none found.
func (m *GetNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetNotificationResponseValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetNotificationResponseValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetNotificationResponseValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetNotificationResponseMultiError(errors)
	}

	return nil
}

// GetNotificationResponseMultiError is an error wrapping multiple validation
// errors returned by GetNotificationResponse.ValidateAll() if the designated
// constraints aren't met.
type GetNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetNotificationResponseMultiError) AllErrors() []error { return m }

// GetNotificationResponseValidationError is the validation error returned by
// GetNotificationResponse.Validate if the designated constraints aren't met.
type GetNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetNotificationResponseValidationError) ErrorName() string {
	return "GetNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetNotificationResponseValidationError{}

// Validate checks the field values on GetNotificationsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetNotificationsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetNotificationsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetNotificationsRequestMultiError, or nil if none found.
func (m *GetNotificationsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetNotificationsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFilter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetNotificationsRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetNotificationsRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFilter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetNotificationsRequestValidationError{
				field:  "Filter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetNotificationsRequestMultiError(errors)
	}

	return nil
}

// GetNotificationsRequestMultiError is an error wrapping multiple validation
// errors returned by GetNotificationsRequest.ValidateAll() if the designated
// constraints aren't met.
type GetNotificationsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetNotificationsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetNotificationsRequestMultiError) AllErrors() []error { return m }

// GetNotificationsRequestValidationError is the validation error returned by
// GetNotificationsRequest.Validate if the designated constraints aren't met.
type GetNotificationsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetNotificationsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetNotificationsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetNotificationsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetNotificationsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetNotificationsRequestValidationError) ErrorName() string {
	return "GetNotificationsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetNotificationsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetNotificationsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetNotificationsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetNotificationsRequestValidationError{}

// Validate checks the field values on GetNotificationsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetNotificationsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetNotificationsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetNotificationsResponseMultiError, or nil if none found.
func (m *GetNotificationsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetNotificationsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetNotifications() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, GetNotificationsResponseValidationError{
						field:  fmt.Sprintf("Notifications[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, GetNotificationsResponseValidationError{
						field:  fmt.Sprintf("Notifications[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return GetNotificationsResponseValidationError{
					field:  fmt.Sprintf("Notifications[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return GetNotificationsResponseMultiError(errors)
	}

	return nil
}

// GetNotificationsResponseMultiError is an error wrapping multiple validation
// errors returned by GetNotificationsResponse.ValidateAll() if the designated
// constraints aren't met.
type GetNotificationsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetNotificationsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetNotificationsResponseMultiError) AllErrors() []error { return m }

// GetNotificationsResponseValidationError is the validation error returned by
// GetNotificationsResponse.Validate if the designated constraints aren't met.
type GetNotificationsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetNotificationsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetNotificationsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetNotificationsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetNotificationsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetNotificationsResponseValidationError) ErrorName() string {
	return "GetNotificationsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetNotificationsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetNotificationsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetNotificationsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetNotificationsResponseValidationError{}

// Validate checks the field values on CreateNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateNotificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateNotificationRequestMultiError, or nil if none found.
func (m *CreateNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreateNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreateNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreateNotificationRequestValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CreateNotificationRequestMultiError(errors)
	}

	return nil
}

// CreateNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by CreateNotificationRequest.ValidateAll() if the
// designated constraints aren't met.
type CreateNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateNotificationRequestMultiError) AllErrors() []error { return m }

// CreateNotificationRequestValidationError is the validation error returned by
// CreateNotificationRequest.Validate if the designated constraints aren't met.
type CreateNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateNotificationRequestValidationError) ErrorName() string {
	return "CreateNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateNotificationRequestValidationError{}

// Validate checks the field values on CreateNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateNotificationResponseMultiError, or nil if none found.
func (m *CreateNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Created

	if len(errors) > 0 {
		return CreateNotificationResponseMultiError(errors)
	}

	return nil
}

// CreateNotificationResponseMultiError is an error wrapping multiple
// validation errors returned by CreateNotificationResponse.ValidateAll() if
// the designated constraints aren't met.
type CreateNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateNotificationResponseMultiError) AllErrors() []error { return m }

// CreateNotificationResponseValidationError is the validation error returned
// by CreateNotificationResponse.Validate if the designated constraints aren't met.
type CreateNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateNotificationResponseValidationError) ErrorName() string {
	return "CreateNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateNotificationResponseValidationError{}

// Validate checks the field values on UpdateNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateNotificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateNotificationRequestMultiError, or nil if none found.
func (m *UpdateNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateNotificationRequestValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateNotificationRequestMultiError(errors)
	}

	return nil
}

// UpdateNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by UpdateNotificationRequest.ValidateAll() if the
// designated constraints aren't met.
type UpdateNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateNotificationRequestMultiError) AllErrors() []error { return m }

// UpdateNotificationRequestValidationError is the validation error returned by
// UpdateNotificationRequest.Validate if the designated constraints aren't met.
type UpdateNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateNotificationRequestValidationError) ErrorName() string {
	return "UpdateNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateNotificationRequestValidationError{}

// Validate checks the field values on UpdateNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateNotificationResponseMultiError, or nil if none found.
func (m *UpdateNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Updated

	if len(errors) > 0 {
		return UpdateNotificationResponseMultiError(errors)
	}

	return nil
}

// UpdateNotificationResponseMultiError is an error wrapping multiple
// validation errors returned by UpdateNotificationResponse.ValidateAll() if
// the designated constraints aren't met.
type UpdateNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateNotificationResponseMultiError) AllErrors() []error { return m }

// UpdateNotificationResponseValidationError is the validation error returned
// by UpdateNotificationResponse.Validate if the designated constraints aren't met.
type UpdateNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateNotificationResponseValidationError) ErrorName() string {
	return "UpdateNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateNotificationResponseValidationError{}

// Validate checks the field values on DeleteNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteNotificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteNotificationRequestMultiError, or nil if none found.
func (m *DeleteNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DeleteNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DeleteNotificationRequestValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DeleteNotificationRequestValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return DeleteNotificationRequestMultiError(errors)
	}

	return nil
}

// DeleteNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by DeleteNotificationRequest.ValidateAll() if the
// designated constraints aren't met.
type DeleteNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteNotificationRequestMultiError) AllErrors() []error { return m }

// DeleteNotificationRequestValidationError is the validation error returned by
// DeleteNotificationRequest.Validate if the designated constraints aren't met.
type DeleteNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteNotificationRequestValidationError) ErrorName() string {
	return "DeleteNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteNotificationRequestValidationError{}

// Validate checks the field values on DeleteNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeleteNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeleteNotificationResponseMultiError, or nil if none found.
func (m *DeleteNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Deleted

	if len(errors) > 0 {
		return DeleteNotificationResponseMultiError(errors)
	}

	return nil
}

// DeleteNotificationResponseMultiError is an error wrapping multiple
// validation errors returned by DeleteNotificationResponse.ValidateAll() if
// the designated constraints aren't met.
type DeleteNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteNotificationResponseMultiError) AllErrors() []error { return m }

// DeleteNotificationResponseValidationError is the validation error returned
// by DeleteNotificationResponse.Validate if the designated constraints aren't met.
type DeleteNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteNotificationResponseValidationError) ErrorName() string {
	return "DeleteNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DeleteNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteNotificationResponseValidationError{}

// Validate checks the field values on FindNotificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindNotificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindNotificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindNotificationRequestMultiError, or nil if none found.
func (m *FindNotificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *FindNotificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFilter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, FindNotificationRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, FindNotificationRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFilter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return FindNotificationRequestValidationError{
				field:  "Filter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return FindNotificationRequestMultiError(errors)
	}

	return nil
}

// FindNotificationRequestMultiError is an error wrapping multiple validation
// errors returned by FindNotificationRequest.ValidateAll() if the designated
// constraints aren't met.
type FindNotificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindNotificationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindNotificationRequestMultiError) AllErrors() []error { return m }

// FindNotificationRequestValidationError is the validation error returned by
// FindNotificationRequest.Validate if the designated constraints aren't met.
type FindNotificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindNotificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindNotificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindNotificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindNotificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindNotificationRequestValidationError) ErrorName() string {
	return "FindNotificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e FindNotificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindNotificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindNotificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindNotificationRequestValidationError{}

// Validate checks the field values on FindNotificationResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindNotificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindNotificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindNotificationResponseMultiError, or nil if none found.
func (m *FindNotificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *FindNotificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetNotification()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, FindNotificationResponseValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, FindNotificationResponseValidationError{
					field:  "Notification",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetNotification()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return FindNotificationResponseValidationError{
				field:  "Notification",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return FindNotificationResponseMultiError(errors)
	}

	return nil
}

// FindNotificationResponseMultiError is an error wrapping multiple validation
// errors returned by FindNotificationResponse.ValidateAll() if the designated
// constraints aren't met.
type FindNotificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindNotificationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindNotificationResponseMultiError) AllErrors() []error { return m }

// FindNotificationResponseValidationError is the validation error returned by
// FindNotificationResponse.Validate if the designated constraints aren't met.
type FindNotificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindNotificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindNotificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindNotificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindNotificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindNotificationResponseValidationError) ErrorName() string {
	return "FindNotificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e FindNotificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindNotificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindNotificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindNotificationResponseValidationError{}

// Validate checks the field values on FindNotificationsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindNotificationsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindNotificationsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindNotificationsRequestMultiError, or nil if none found.
func (m *FindNotificationsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *FindNotificationsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFilter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, FindNotificationsRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, FindNotificationsRequestValidationError{
					field:  "Filter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFilter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return FindNotificationsRequestValidationError{
				field:  "Filter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return FindNotificationsRequestMultiError(errors)
	}

	return nil
}

// FindNotificationsRequestMultiError is an error wrapping multiple validation
// errors returned by FindNotificationsRequest.ValidateAll() if the designated
// constraints aren't met.
type FindNotificationsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindNotificationsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindNotificationsRequestMultiError) AllErrors() []error { return m }

// FindNotificationsRequestValidationError is the validation error returned by
// FindNotificationsRequest.Validate if the designated constraints aren't met.
type FindNotificationsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindNotificationsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindNotificationsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindNotificationsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindNotificationsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindNotificationsRequestValidationError) ErrorName() string {
	return "FindNotificationsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e FindNotificationsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindNotificationsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindNotificationsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindNotificationsRequestValidationError{}

// Validate checks the field values on FindNotificationsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindNotificationsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindNotificationsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindNotificationsResponseMultiError, or nil if none found.
func (m *FindNotificationsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *FindNotificationsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetNotifications() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, FindNotificationsResponseValidationError{
						field:  fmt.Sprintf("Notifications[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, FindNotificationsResponseValidationError{
						field:  fmt.Sprintf("Notifications[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return FindNotificationsResponseValidationError{
					field:  fmt.Sprintf("Notifications[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return FindNotificationsResponseMultiError(errors)
	}

	return nil
}

// FindNotificationsResponseMultiError is an error wrapping multiple validation
// errors returned by FindNotificationsResponse.ValidateAll() if the
// designated constraints aren't met.
type FindNotificationsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindNotificationsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindNotificationsResponseMultiError) AllErrors() []error { return m }

// FindNotificationsResponseValidationError is the validation error returned by
// FindNotificationsResponse.Validate if the designated constraints aren't met.
type FindNotificationsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindNotificationsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindNotificationsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindNotificationsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindNotificationsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindNotificationsResponseValidationError) ErrorName() string {
	return "FindNotificationsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e FindNotificationsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindNotificationsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindNotificationsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindNotificationsResponseValidationError{}

// Validate checks the field values on Notification_Action with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *Notification_Action) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Notification_Action with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// Notification_ActionMultiError, or nil if none found.
func (m *Notification_Action) ValidateAll() error {
	return m.validate(true)
}

func (m *Notification_Action) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	switch v := m.Action.(type) {
	case *Notification_Action_RegisterUser_:
		if v == nil {
			err := Notification_ActionValidationError{
				field:  "Action",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetRegisterUser()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, Notification_ActionValidationError{
						field:  "RegisterUser",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, Notification_ActionValidationError{
						field:  "RegisterUser",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetRegisterUser()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return Notification_ActionValidationError{
					field:  "RegisterUser",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return Notification_ActionMultiError(errors)
	}

	return nil
}

// Notification_ActionMultiError is an error wrapping multiple validation
// errors returned by Notification_Action.ValidateAll() if the designated
// constraints aren't met.
type Notification_ActionMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m Notification_ActionMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m Notification_ActionMultiError) AllErrors() []error { return m }

// Notification_ActionValidationError is the validation error returned by
// Notification_Action.Validate if the designated constraints aren't met.
type Notification_ActionValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e Notification_ActionValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e Notification_ActionValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e Notification_ActionValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e Notification_ActionValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e Notification_ActionValidationError) ErrorName() string {
	return "Notification_ActionValidationError"
}

// Error satisfies the builtin error interface
func (e Notification_ActionValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotification_Action.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = Notification_ActionValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = Notification_ActionValidationError{}

// Validate checks the field values on Notification_Action_RegisterUser with
// the rules defined in the proto definition for this message. If any rules
// are violated, the first error encountered is returned, or nil if there are
// no violations.
func (m *Notification_Action_RegisterUser) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Notification_Action_RegisterUser with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// Notification_Action_RegisterUserMultiError, or nil if none found.
func (m *Notification_Action_RegisterUser) ValidateAll() error {
	return m.validate(true)
}

func (m *Notification_Action_RegisterUser) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return Notification_Action_RegisterUserMultiError(errors)
	}

	return nil
}

// Notification_Action_RegisterUserMultiError is an error wrapping multiple
// validation errors returned by
// Notification_Action_RegisterUser.ValidateAll() if the designated
// constraints aren't met.
type Notification_Action_RegisterUserMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m Notification_Action_RegisterUserMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m Notification_Action_RegisterUserMultiError) AllErrors() []error { return m }

// Notification_Action_RegisterUserValidationError is the validation error
// returned by Notification_Action_RegisterUser.Validate if the designated
// constraints aren't met.
type Notification_Action_RegisterUserValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e Notification_Action_RegisterUserValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e Notification_Action_RegisterUserValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e Notification_Action_RegisterUserValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e Notification_Action_RegisterUserValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e Notification_Action_RegisterUserValidationError) ErrorName() string {
	return "Notification_Action_RegisterUserValidationError"
}

// Error satisfies the builtin error interface
func (e Notification_Action_RegisterUserValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNotification_Action_RegisterUser.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = Notification_Action_RegisterUserValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = Notification_Action_RegisterUserValidationError{}
